### ISSUE TITLE
Unit tests for middleware

### ISSUE THREAD

**Issue by SergKanz** [Microsoft/ApplicationInsights-aspnetv5#62]
I've been trying to create a unit test for middleware. Following [stackoverflow](http://stackoverflow.com/questions/29561277/testing-asp-net-5-vnext-middleware-from-a-testserver) example I wrote this code, but it fails to construct middleware as it cannot find the ```TelemetryClient```. Any ideas? 

```
            sentItems.Clear();
            var configuration = new TelemetryConfiguration()
            {
                TelemetryChannel = new FakeTelemetryChannel() { OnSend = item => this.sentItems.Add(item) },
                InstrumentationKey = "testestkey"
            };
            var client = new TelemetryClient(configuration);

            var services = new ServiceCollection();
            services.AddInstance<TelemetryClient>(client);

            services.AddInstance<IApplicationEnvironment>(CallContextServiceLocator.Locator.ServiceProvider.GetService<IApplicationEnvironment>());
            services.AddInstance<IServiceManifest>(CallContextServiceLocator.Locator.ServiceProvider.GetService<IServiceManifest>());

            var server = TestServer.Create(services.BuildServiceProvider(), (IApplicationBuilder app) =>
            {
                app.UseMiddleware<RequestTrackingMiddleware>();
                app.Use(next =>
                {
                    return async context =>
                    {
                        await context.Response.WriteAsync("Hello, world!");
                    };
                });
            });

            using (server)
            {
                var response = await server.CreateClient().GetAsync("/");
                Assert.Equal(HttpStatusCode.OK, response.StatusCode);
            }
```

**Comment by SergKanz** [comment]
@olegsych @upendras - this may be helpful for testing bugs that you took. However I'm not sure why it is not working

**Comment by olegsych** [comment]
Thanks for pointing this out, @SergKanz . `TestServer` should allow us to replace our own `InProcessServer` from the FunctionalTestUtils. I'll try to get it running with the `TelemetryClient` service. 

P.S. This is definitely not a _unit_ test. I hope we don't have to write a lot of tests like it.

**Comment by SergKanz** [comment]
My thinking was that unit testing of middleware will require too much of mocking. That's why I was thinking of TestServer. Never tried though. Will be interesting to see how they will look like and compare with using of TestServer

**Comment by olegsych** [comment]
Assuming that you mean mocking things like `IApplicationEnvironment` and `IServiceManifest`, I would agree. However, this is _integration_ testing, because the test uses the `RequestTrackingMiddleware` component with other components, like the `TestServer` itself. Integration tests tend to be more [obscure](http://xunitpatterns.com/Obscure%20Test.html) than unit tests, because they require complex setup of framework dependencies that aren't related to the system under test and more [fragile](http://xunitpatterns.com/Fragile%20Test.html) because they break when framework dependencies change.

A _unit_ test for a middleware component, like the [ExceptionTrackingMiddlewareTest](https://github.com/Microsoft/ApplicationInsights-aspnetv5/blob/develop/test/Microsoft.ApplicationInsights.AspNet.Tests/ExceptionTrackingMiddlewareTest.cs) does not need to worry about unrelated dependencies because it uses a [minimal fixture](http://xunitpatterns.com/Minimal%20Fixture.html) to test the SUT directly, without the framework infrastructure and the DI container in particular. 

And now that I stopped preaching, :sleepy: I actually got your code snippet running :smile:!
```c#
var server = TestServer.Create(
    (IApplicationBuilder app) =>
    {
        app.UseMiddleware<RequestTrackingMiddleware>();
        app.Use((RequestDelegate next) =>
        {
            return async (HttpContext context) =>
            {
                await context.Response.Body.WriteAsync(new byte[0], 0, 0);
            };
        });
    },
    (IServiceCollection services) =>
    {
        services.AddInstance(new TelemetryClient(new TelemetryConfiguration()));
        services.AddInstance(new RequestTelemetry());
    });

using (server)
{
    var response = await server.CreateClient().GetAsync("/");
    Assert.Equal(HttpStatusCode.OK, response.StatusCode);
}
```
